lib/date_time/date_time.c

```diff
@@ -0,0 +1,301 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <date_time.h>
+#include <zephyr.h>
+#include <zephyr/types.h>
+#include <device.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <at_cmd.h>
+#include <time.h>
+#include <errno.h>
+#include <string.h>
+#include <net/sntp.h>
+#include <net/socketutils.h>
+#include <sys/timeutil.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(date_time, CONFIG_DATE_TIME_LOG_LEVEL);
+
+#define AT_CMD_MODEM_DATE_TIME			"AT+CCLK?"
+#define AT_CMD_MODEM_DATE_TIME_RESPONSE_LEN	28
+
+/* The magic number 115 corresponds to the year 2015 which is the default
+ * year given by the modem when the cellular network has not pushed time
+ * the modem.
+ */
+#define MODEM_TIME_DEFAULT 115
+
+#define UIO_IP      "129.240.2.6"
+#define GOOGLE_IP   "216.239.35.0"
+#define GOOGLE_IP_2 "216.239.35.4"
+#define GOOGLE_IP_3 "216.239.35.8"
+#define GOOGLE_IP_4 "216.239.35.12"
+
+#define NTP_DEFAULT_PORT "123"
+
+K_SEM_DEFINE(time_fetch_sem, 0, 1);
+
+static struct k_delayed_work time_work;
+
+static struct time_aux {
+	s64_t date_time_utc;
+	int last_date_time_update;
+} time_aux;
+
+static struct sntp_time sntp_time;
+static bool initial_valid_time;
+
+struct ntp_servers {
+	const char *server;
+	struct addrinfo *addr;
+};
+
+struct ntp_servers servers[] = {
+	{.server = UIO_IP},
+	{.server = GOOGLE_IP},
+	{.server = GOOGLE_IP_2},
+	{.server = GOOGLE_IP_3},
+	{.server = GOOGLE_IP_4}
+};
+
+#if defined(CONFIG_DATE_TIME_MODEM)
+static int time_entries_parse(char *datetime_string, int min, int max)
+{
+	char buf[50];
+
+	memcpy(&buf, &datetime_string[min], max - min + 1);
+
+	buf[max - min + 1] = '\0';
+
+	return strtol(buf, NULL, 10);
+}
+
+static int time_modem_get(void)
+{
+	int err;
+	char buf[AT_CMD_MODEM_DATE_TIME_RESPONSE_LEN + 5];
```

Comment:
Shouldn't it be + 1 for the terminating NULL character? What are the other 4 bytes for?

Reply:
Adjusted the ``AT_CMD_MODEM_DATE_TIME_RESPONSE_LEN`` and set the extended length to 1. I kept getting errors that the response buffer for the at command was not large enough. 5 was the lowest value that did not give me this error. The actual response is:
"+CCLK: \"20/02/25,17:15:02+04\"\r\n\000{". This should now be accounted for.